File: Strand Sort/StrandSort.py
def merge(a, b):

    result = []
    i = j = 0

    while i < len(a) and j < len(b):

        if a[i] < b[j]:
            result.append(a[i])
            i += 1
        else:
            result.append(b[j])
            j += 1

    result.extend(a[i:])
    result.extend(b[j:])

    return result

def strand_sort(arr):

    result = []

    while arr:

        sublist = [arr.pop(0)]

        i = 0

        while i < len(arr):

            if arr[i] >= sublist[-1]:
                sublist.append(arr.pop(i))
            else:
                i += 1

        result = merge(result, sublist)

    return result

arr = [10, 5, 30, 40, 2, 4, 9]

sorted_arr = strand_sort(arr)

print(sorted_arr)

File: Strand Sort/StrandSort.java
import java.util.ArrayList;

class StrandSort {

    static ArrayList<Integer> merge(ArrayList<Integer> a, ArrayList<Integer> b) {

        ArrayList<Integer> result = new ArrayList<>();

        int i = 0, j = 0;

        while(i < a.size() && j < b.size()) {

            if(a.get(i) < b.get(j))
                result.add(a.get(i++));
            else
                result.add(b.get(j++));
        }

        while(i < a.size())
            result.add(a.get(i++));

        while(j < b.size())
            result.add(b.get(j++));

        return result;
    }

    static void strandSort(ArrayList<Integer> arr) {

        ArrayList<Integer> result = new ArrayList<>();

        while(!arr.isEmpty()) {

            ArrayList<Integer> sublist = new ArrayList<>();
            sublist.add(arr.remove(0));

            for(int i = 0; i < arr.size(); ) {

                if(arr.get(i) >= sublist.get(sublist.size() - 1))
                    sublist.add(arr.remove(i));
                else
                    i++;
            }

            result = merge(result, sublist);
        }

        arr.addAll(result);
    }

    public static void main(String[] args) {

        ArrayList<Integer> arr = new ArrayList<>();

        arr.add(10);
        arr.add(5);
        arr.add(30);
        arr.add(40);
        arr.add(2);
        arr.add(4);
        arr.add(9);

        strandSort(arr);

        for(int num : arr)
            System.out.print(num + " ");
    }
}
